cadc-inventory-server/src/main/java/org/opencadc/inventory/transfer/ProtocolsGenerator.java

```diff
@@ -486,52 +508,50 @@ private List<Protocol> doPushTo(URI artifactURI, Transfer transfer, String authT
-                        Interface iface = filesCap.findInterface(sec);
-                        log.debug("PUT: " + storageSite + " proto: " + proto + " iface: " + iface);
-                        if (iface != null) {
-                            URL baseURL = iface.getAccessURL().getURL();
-                            //log.debug("base url for site " + storageSite.getResourceID() + ": " + baseURL);
-                            if (protocolCompat(proto, baseURL)) {
-                                // // no plain anon URL for put: !anon or anon+token
-                                boolean gen = (!anon || (anon && authToken != null));
-                                if (gen) {
-                                    StringBuilder sb = new StringBuilder();
-                                    sb.append(baseURL.toExternalForm()).append("/");
-                                    if (authToken != null && anon) {
-                                        sb.append(authToken).append("/");
-                                    }
-                                    sb.append(artifactURI.toASCIIString());
-                                    Protocol p = new Protocol(proto.getUri());
-                                    if (transfer.version == VOS.VOSPACE_21) {
-                                        p.setSecurityMethod(proto.getSecurityMethod());
-                                    }
-                                    p.setEndpoint(sb.toString());
-                                    protos.add(p);
-                                    log.debug("added: " + p);
+                    log.debug("pushTo: " + storageSite + " proto: " + proto);
+                    // less generic request for service that implements
+                    // HACK: this is filesCap specific in here
+                    if (proto.getUri().equals(filesCap.getStandardID())) {
+                        Protocol p = new Protocol(proto.getUri());
+                        p.setEndpoint(storageSite.getResourceID().toASCIIString());
+                        protos.add(p);
+                    }
+                    URI sec = proto.getSecurityMethod();
+                    if (sec == null) {
+                        sec = Standards.SECURITY_METHOD_ANON;
+                    }
+                    boolean anon = Standards.SECURITY_METHOD_ANON.equals(sec);
+                    Interface iface = filesCap.findInterface(sec);
+                    log.debug("pushTo: " + storageSite + " proto: " + proto + " iface: " + iface);
+                    if (iface != null) {
+                        URL baseURL = iface.getAccessURL().getURL();
+                        //log.debug("base url for site " + storageSite.getResourceID() + ": " + baseURL);
+                        if (protocolCompat(proto, baseURL)) {
+                            // // no plain anon URL for put: !anon or anon+token
+                            boolean gen = (!anon || (anon && authToken != null));
```

Comment:
From what I'm seeing, the `pushTo` and `pullFrom` loops are pretty similar except this and the file name override. Would it be possible (or worth it) to avoid the code duplication?

Reply:
maybe, but iirc there were enough subtle differences that made that more complex. I think you originally wrote it and convinced me of that, although now there's plenty of `git blame` to go around :-)